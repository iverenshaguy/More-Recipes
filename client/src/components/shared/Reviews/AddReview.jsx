import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Form from '../../shared/Forms';
import './AddReview.scss';

/**
 * @exports
 * Returns Review Wrapper
 * @param {object} props
 * @returns {JSX} RecipeWrapper
 */
const AddReview = (props) => {
  const meta = {
    title: 'Review this Recipe',
    btnText: 'Add Your Review'
  };

  const reviewFormClass = props.showReviewForm ? 'my-3' : 'd-none';

  return (
    <div className={reviewFormClass} id="review-wrapper">
      <div id="review-div">
        <div className="card card-block px-md-5 px-sm-3 py-3">
          <Form {...props} type="review" id={props.id} meta={meta} />
        </div>
      </div>
    </div>
  );
};

AddReview.propTypes = {
  id: PropTypes.number.isRequired,
  showReviewForm: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  showReviewForm: state.ui.reviewForm.isOpen,
  id: state.singleRecipe.recipe.item.recipeItem.id,
  submitError: state.singleRecipe.recipeReviews.error,
  submitting: state.singleRecipe.recipeReviews.reviewing
});

export { AddReview as AddReviewComponent };

export default connect(mapStateToProps)(AddReview);
