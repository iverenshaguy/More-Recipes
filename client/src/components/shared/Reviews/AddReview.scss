#review-wrapper {
  h4,
  hr {
    display: none;
  }
}
