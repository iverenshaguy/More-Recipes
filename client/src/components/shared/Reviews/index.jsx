import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ReactStars from 'react-stars';
import Pagination from '../Pagination';
import { multiReviewPropTypes } from '../../../helpers/proptypes';
import { fetchReviews } from '../../../actions/recipeReviews';
import MiniPreLoader from '../../shared/PreLoader/MiniPreLoader';
import getNoResultText from '../../../helpers/getNoResultText';
import AddReview from './AddReview';

/**
 * Returns Reviews Wrapper
 * @exports
 * @extends Component
 * @class Reviews
 * @param {array} reviews
 * @returns {JSX} Reviews
 */
class Reviews extends Component {
  static propTypes = {
    error: PropTypes.string,
    recipeId: PropTypes.number.isRequired,
    dispatch: PropTypes.func.isRequired,
    reviewing: PropTypes.bool.isRequired,
    ...multiReviewPropTypes
  }

  static defaultProps = {
    error: null
  }

  /**
   * @constructor
   * @memberof Reviews
   * @returns {nothing} - Returns nothing
   */
  constructor() {
    super();

    this.state = {
      currentPage: 1,
      limit: 10
    };

    this.getReviews = this.getReviews.bind(this);
    this.handlePageChange = this.handlePageChange.bind(this);
  }

  /**
   * @memberof Reviews
   * @returns {noting} Returns nothing
   */
  componentWillMount() {
    this.getReviews();
  }

  /**
   * @memberof Reviews
   * @returns {noting} Returns nothing
   */
  getReviews() {
    const { currentPage, limit } = this.state;
    this.props.dispatch(fetchReviews(this.props.recipeId, currentPage, limit));
  }

  /**
   * @memberof Reviews
   * @param {number} page
   * @returns {state} home
   */
  handlePageChange(page) {
    this.setState({
      currentPage: page
    }, () => this.getReviews());
  }

  /**
   * @memberof Reviews
   * @returns {JSX} Reviews
   */
  render() {
    const {
      reviews, reviewing, error, metadata
    } = this.props;

    return (
      <div className="py-5 p-xs-1 p-md-5" id="reviews">
        <div className="col-12">
          <h4>Reviews</h4>
          <hr />
        </div>
        {reviewing && <div className="text-center"><MiniPreLoader /></div>}
        {!reviewing && !!reviews && reviews.length === 0 && <div className="text-center no-reviews">{getNoResultText('REVIEWS RESULT')}</div>}
        {error && <div className="text-center error">Something happened, please retry.</div>}
        {!!reviews && reviews.length !== 0 && reviews.sort((a, b) => b.id - a.id).map(review => (
          <div className="px-5 pb-3 review-heading" key={review.id}>
            <div className="row">
              <div className="col-4 rounded-circle text-left">
                <img src={review.User.profilePic} alt="user" className="review-picture" />
              </div>
              <div className="col-8 mr-0 text-right">
                <p className="mb-0">{review.User.username}</p>
              </div>
            </div>
            <hr style={{ marginTop: '0.5em', marginBottom: '0.5em' }} />
            <ReactStars value={+review.rating} size={30} color2="#f1c40f" half edit={false} />
            <p className="text-justify">{review.comment}</p>
          </div>))}
        {reviews.length !== 0 && <Pagination
          items={reviews}
          onPageChange={this.handlePageChange}
          metadata={metadata}
        />}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  reviews: state.singleRecipe.recipeReviews.reviews,
  metadata: state.singleRecipe.recipeReviews.metadata,
  error: state.singleRecipe.recipeReviews.error,
  reviewing: state.singleRecipe.recipeReviews.reviewing
});

export { AddReview, Reviews as ReviewsComponent };
export default connect(mapStateToProps)(Reviews);
