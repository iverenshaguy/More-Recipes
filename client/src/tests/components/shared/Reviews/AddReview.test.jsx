import React from 'react';
import { AddReviewComponent as AddReview } from '../../../../components/shared/Reviews/AddReview';

const setup = () => {
  const props = {
    id: 2,
    submitting: false,
    submitError: null,
    showReviewForm: true
  };

  const shallowWrapper = shallow(<AddReview {...props} />);

  return { props, shallowWrapper };
};

describe('AddReview', () => {
  it('renders correctly', () => {
    const { shallowWrapper } = setup();
    expect(toJson(shallowWrapper)).toMatchSnapshot();
    expect(shallowWrapper.find('.d-none').length).toBeFalsy();
  });

  it('renders correctly when showReviewForm is false', () => {
    const { props } = setup();
    const shallowWrapper = shallow(<AddReview {...props} showReviewForm={false} />);
    expect(shallowWrapper.find('.d-none').length).toBeTruthy();
  });
});
