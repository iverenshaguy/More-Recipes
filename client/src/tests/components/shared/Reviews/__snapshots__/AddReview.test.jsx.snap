// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReview renders correctly 1`] = `
<div
  className="my-3"
  id="review-wrapper"
>
  <div
    id="review-div"
  >
    <div
      className="card card-block px-md-5 px-sm-3 py-3"
    >
      <Connect(Form)
        id={2}
        meta={
          Object {
            "btnText": "Add Your Review",
            "title": "Review this Recipe",
          }
        }
        showReviewForm={true}
        submitError={null}
        submitting={false}
        type="review"
      />
    </div>
  </div>
</div>
`;
